pkg/tbtc/chain_test.go

```diff
@@ -129,74 +143,291 @@ func (lc *localChain) OnDKGResultSubmitted(
+	groupSelectionResult *GroupSelectionResult,
 ) error {
 	lc.dkgResultSubmissionHandlersMutex.Lock()
 	defer lc.dkgResultSubmissionHandlersMutex.Unlock()
 
-	lc.resultSubmissionMutex.Lock()
-	defer lc.resultSubmissionMutex.Unlock()
+	lc.dkgMutex.Lock()
+	defer lc.dkgMutex.Unlock()
+
+	if lc.dkgState != AwaitingResult {
+		return fmt.Errorf("not awaiting DKG result")
+	}
 
 	blockNumber, err := lc.blockCounter.CurrentBlock()
 	if err != nil {
 		return fmt.Errorf("failed to get the current block")
 	}
 
-	groupPublicKeyBytes, err := result.GroupPublicKeyBytes()
+	groupPublicKeyBytes, err := tecdsaDkgResult.GroupPublicKeyBytes()
 	if err != nil {
 		return fmt.Errorf(
 			"failed to extract group public key bytes from the result [%v]",
 			err,
 		)
 	}
 
+	signingMembersIndexes := make([]group.MemberIndex, 0)
+	signaturesConcatenation := make([]byte, 0)
+	for memberIndex, signature := range signatures {
+		signingMembersIndexes = append(signingMembersIndexes, memberIndex)
+		signaturesConcatenation = append(signaturesConcatenation, signature...)
+	}
+
+	operatingMembersIndexes := tecdsaDkgResult.Group.OperatingMemberIndexes()
+	operatingOperatorsIDsBytes := make([]byte, 0)
+	for _, operatingMemberID := range operatingMembersIndexes {
+		operatorIDBytes := make([]byte, 4)
+		operatorID := groupSelectionResult.OperatorsIDs[operatingMemberID-1]
+		binary.BigEndian.PutUint32(operatorIDBytes, operatorID)
+
+		operatingOperatorsIDsBytes = append(
+			operatingOperatorsIDsBytes,
+			operatorIDBytes...,
+		)
+	}
+
+	result := &DKGChainResult{
+		SubmitterMemberIndex:     memberIndex,
+		GroupPublicKey:           groupPublicKeyBytes,
+		MisbehavedMembersIndexes: tecdsaDkgResult.MisbehavedMembersIndexes(),
+		Signatures:               signaturesConcatenation,
+		SigningMembersIndexes:    signingMembersIndexes,
+		Members:                  groupSelectionResult.OperatorsIDs,
+		MembersHash:              sha3.Sum256(operatingOperatorsIDsBytes),
+	}
+
+	resultHash := computeTestDkgResultHash(result)
+
 	for _, handler := range lc.dkgResultSubmissionHandlers {
 		handler(&DKGResultSubmittedEvent{
-			MemberIndex:         uint32(memberIndex),
-			GroupPublicKeyBytes: groupPublicKeyBytes,
-			Misbehaved:          result.MisbehavedMembersIndexes(),
-			BlockNumber:         blockNumber,
+			Seed:        nil,
+			ResultHash:  resultHash,
+			Result:      result,
+			BlockNumber: blockNumber,
 		})
 	}
 
-	lc.activeWallet = groupPublicKeyBytes
-	lc.resultSubmitterIndex = memberIndex
+	lc.dkgState = Challenge
+	lc.dkgResult = result
+	lc.dkgResultValid = true
 
 	return nil
 }
 
-// GetDKGState returns the current state of the DKG procedure.
 func (lc *localChain) GetDKGState() (DKGState, error) {
-	return AwaitingResult, nil
+	lc.dkgMutex.Lock()
+	defer lc.dkgMutex.Unlock()
+
+	return lc.dkgState, nil
 }
 
-// CalculateDKGResultHash calculates 256-bit hash of DKG result using SHA3-256
-// hashing algorithm.
 func (lc *localChain) CalculateDKGResultHash(
```

Comment:
I had similar confusion when looking at this function vs `computeTestDkgResultHash` at the bottom of the file as when writing [this](https://github.com/keep-network/keep-core/pull/3427#discussion_r1052246527) comment. 

In the contract, we use the same `Result` struct everywhere: to submit the result, to challenge the result, and to approve the result. The hash is computed based on the `Result` struct.

In `tbtc.go`, when we submit the result we use `*dkg.Result` plus some additional parameters, and when we approve or challenge the result we use `*DKGChainResult`. The hash is computed based on the `*dkg.Result` struct (which I am not sure is entirely correct - see [the other comment](https://github.com/keep-network/keep-core/pull/3427#discussion_r1052261813) please).

In the test chain implementation, we follow the same approach as `tbtc.go`. The hash is computed based on the `*dkg.Result` struct, `*DKGChainResult` is used to approve and challenge a result. Internally, the result is stored as `*DKGChainResult`

Depending on the result of the discussion from two other comments referenced, I think we should clarify this code as well. 

I think we can:
1. Provide a function that will transform `*dkg.Result` to `DKGChainResult` and compute hash from `DKGChainResult`. Also, use `DKGChainResult` to `SubmitDKGResult`. In local chain implementation, we would have only one function for computing a hash.
2. Rename `DKGChainResult` to `DKGChainSubmittedResult` to clearly indicate it's an on-chain implementation of a result that is already submitted and can only be challenged or approved, improve the documentation to make it clear. But then... what to do about hashes, why to compute hashes in two different ways if it's not what the chain expects 🤔 .

Reply:
Please see https://github.com/keep-network/keep-core/pull/3427#discussion_r1053347458